varinfo/generate_umm_var.py

```diff
@@ -0,0 +1,92 @@
+    * Parse the variable metadata from the granule using VarInfoFromNetCDF4.
+    * Generate UMM-Var JSON as a dictionary for each identified variable.
+    * (Optional) Publish each UMM-Var entry to the selected CMR environment.
+
+'''
+from tempfile import TemporaryDirectory
+from typing import Dict, List, Union
+import re
+
+from cmr import CMR_UAT
+
+from varinfo import VarInfoFromNetCDF4
+from varinfo.cmr_search import (CmrEnvType, download_granule, get_granule_link,
+                                get_granules)
+from varinfo.umm_var import get_all_umm_var, publish_all_umm_var
+
+
+# Custom return type: either a list of UMM-Var JSON (a list of dictionaries),
+# or a list of strings (either concept IDs or error strings).
+UmmVarReturnType = List[Union[Dict, str]]
+
+
+def generate_collection_umm_var(collection_concept_id: str,
+                                auth_header: str,
+                                cmr_env: CmrEnvType = CMR_UAT,
+                                publish: bool = False) -> UmmVarReturnType:
+    """ Run all the of the functions for downloading and publishing
+        a UMM-Var entry to CMR given:
+
+        * collection_concept_id: Concept ID for collection that variables have
+          been generated for.
+        * cmr_env: URLs for CMR environments (OPS, UAT, and SIT)
+        * auth_header: Authorization HTTP header, containing either:
+          - A LaunchPad token: 'Authorization: <token>'
+          - An EDL bearer token: 'Authorization: Bearer <token>'
+        * publish: Optional argument determining whether to publish the
+          generated UMM-Var records to the indicated CMR instance. Defaults to
+          False.
+
+        Note - if attempting to publish to CMR, a LaunchPad token must be used.
+
+    """
+    granule_response = get_granules(collection_concept_id, cmr_env=cmr_env,
+                                    auth_header=auth_header)
+
+    # Get the data download URL for the most recent granule (NetCDF-4 file)
+    granule_link = get_granule_link(granule_response)
+
+    with TemporaryDirectory() as temp_dir:
+        # Download file to runtime environment
+        local_granule = download_granule(granule_link,
+                                         auth_header,
+                                         out_directory=temp_dir)
+
+        # Parse the granule with VarInfo to map all variables and relations:
+        var_info = VarInfoFromNetCDF4(local_granule)
+
+        # Generate all the UMM-Var records:
+        all_umm_var_records = get_all_umm_var(var_info)
+
+    if publish:
+        # Publish to CMR and construct an output object that is a list of
+        # strings. These strings will be either variable concept IDs or
+        # error messages returned from CMR.
+        publication_response = publish_all_umm_var(collection_concept_id,
+                                                   all_umm_var_records,
+                                                   auth_header, cmr_env)
+
```

Comment:
So for a collection if we have one variable fail ingesting and one pass this is what it looks like: 
`[
  {'/TROPPV': "['#: required key [LongName] not found']", 
  {'/TROPQ': 'V1259791548-EEDTEST'}
  ]`
And can you add whatever it is in the comments like `# e.g. ...`

Reply:
Close, but the output for a response with an error is a bit different. It would look like:

```
[ {'/TROPPV': '#: required key [LongName] not found', {'/TROPQ': 'V1259791548-EEDTEST'} ]
```

If `/TROPPV` returned multiple errors, those get combined.

E.g.: CMR response JSON:

```
{
  "errors": ["error 1", "error 2"]
  ...
}
```

Combined output from `publish_all_umm_var`:

```
[{'/TROPPV': 'error 1\n  error 2', {'/TROPQ': 'V1259791548-EEDTEST'} ]
```

- [ ] I'll review the comments and documentation string for `publish_umm_var` to try better describe that output.